leakcanary-object-watcher/src/main/java/leakcanary/ObjectWatcher.kt

```diff
@@ -134,11 +134,10 @@ class ObjectWatcher constructor(
     val watchUptimeMillis = clock.uptimeMillis()
     val reference =
       KeyedWeakReference(watchedObject, key, name, watchUptimeMillis, queue)
-    SharkLog.d(
-        "Watching %s with key %s",
-        ((if (watchedObject is Class<*>) watchedObject.toString() else "instance of ${watchedObject.javaClass.name}") + if (name.isNotEmpty()) " named $name" else ""),
-        key
-    )
+    val nameForLogging =
+      (if (watchedObject is Class<*>) watchedObject.toString() else "instance of ${watchedObject.javaClass.name}") +
+      (if (name.isNotEmpty()) " named $name" else "")
+    SharkLog.d("Watching $nameForLogging with key $key")
```

Comment:
The previous log was bad, this makes it slightly worst :) . Context: the idea with these logs is that they're no-op when there's no logger, and there's no unnecessary string concatenation. Here the log was actually bad, it should have done something like `if (SharkLog.logger != null) {}` (or maybe SharkLog should have a util method) to avoid any object construction. `"Watching $nameForLogging with key $key"` involves string interpolation even when there is no logger.

This is not a major impact, but more of a good practice.

Reply:
Ooh, I see! I'll restructure in a better way later today to prevent this unnecessary allocation.

On a side note: the previous call to `SharkLog.d` was passing the arguments that were later passed into the `String.format()` to replace `%s` placeholders. That internally uses `Matchers` and `Formatter` which most likely have worse performance than using Kotlin's string template (which uses `StringBuilder`)
So I wonder if it would be better to replace the calls like `logger.d("Logging %s", value)` with `logger.d("Logging $value")`

Reply:
It would be more efficient when logger is set, but less efficient when logger is not set.

Maybe instead SharkLog's API should be changed to take an inline lambda that returns the formatted string:

```
SharkLog.d {
"Logging $value"
}
```

inline means no extra cost, you get the kotlin string templates, and if there's no logger then there's no cost.

Reply:
Wouldn't it be more expensive to create an instance of that lambda?

```
//SharkLog.kt:
fun d(lambda:() -> String) {
  val logger = logger ?: return
  logger.d(lambda.invoke())
}

//Test class:
private fun getI() = 6

@Test fun logTest() {
  val i = getI()
  SharkLog.d { "oops$i" }
  SharkLog.d ("oops$i")
}

//Decompiled Java
@Test
public final void logTest() {
   final int i = this.getI();
   SharkLog.INSTANCE.d((Function0)(new Function0() {
      // $FF: synthetic method
      // $FF: bridge method
      public Object invoke() {
         return this.invoke();
      }
      @NotNull
      public final String invoke() {
         return "oops" + i;
      }
   }));
   SharkLog.INSTANCE.d("oops" + i, new Object[0]);
}
private final int getI() {
   return 6;
}
```
Or am I doing something wrong here?

Reply:
Make `d` an inline function and then the lambda gets inlined

Reply:
Aha! Way better now! 
I keep forgetting about this little magic word 😄 

```
@Test
public final void logTest() {
   int i = this.getI();
   SharkLog this_$iv = SharkLog.INSTANCE;
   int $i$f$d = false;
   Logger var10000 = this_$iv.getLogger();
   if (var10000 != null) {
      Logger logger$iv = var10000;
      int var5 = false;
      String var7 = "oops" + i;
      logger$iv.d(var7, new Object[0]);
   }
   SharkLog.INSTANCE.d("oops" + i, new Object[0]);
}
```

Reply:
with that we can get rid of the vararg and lose the `new Object[0]`

Reply:
I will make it as a separate Pull Request, probably tomorrow